pkg/operator/webhook/add.go

```diff
@@ -109,6 +109,27 @@ func GetMutatingWebhookConfiguration(mode, url string) *admissionregistrationv1.
 				MatchPolicy:    &matchPolicy,
 				TimeoutSeconds: ptr.To[int32](10),
 			},
+			{
+				Name:                    "extension.defaulting.operator.gardener.cloud",
+				ClientConfig:            getClientConfig(extension.WebhookPath, mode, url),
+				AdmissionReviewVersions: []string{"v1", "v1beta1"},
+				Rules: []admissionregistrationv1.RuleWithOperations{{
+					Rule: admissionregistrationv1.Rule{
+						APIGroups:   []string{operatorv1alpha1.SchemeGroupVersion.Group},
+						APIVersions: []string{operatorv1alpha1.SchemeGroupVersion.Version},
+						Resources:   []string{"extensions"},
+					},
+					Operations: []admissionregistrationv1.OperationType{
+						admissionregistrationv1.Create,
+						admissionregistrationv1.Update,
+						admissionregistrationv1.Delete,
+					},
+				}},
+				SideEffects:    &sideEffects,
+				FailurePolicy:  &failurePolicy,
+				MatchPolicy:    &matchPolicy,
+				TimeoutSeconds: ptr.To[int32](10),
+			},
```

Comment:
As discussed via Slack yesterday, I still don't know whether it is a good idea to persist our defaults in the object. This means we can not easily change default settings in operator because once they have been written to the spec, we have to use those (since we don't know whether they have been specified by the human or defaulted by us).

You mentioned that without this, we run "implicit upgrades".
- When we update GOP, we need to be sure that the extensions are still compatible with it.

Yes, sure, we have to be more careful now, i.e., when we introduce breaking changes, we have to update the referenced extensions to a version that supports (that's why we first deprecate things to give folks time to adapt, and only "break" later). As usual, extension developers have to adhere. I don't think this is much different than before when you were operating Gardener, is it?

- When we update an extension, we need to be sure that it works well with GOP.

That's even easier, because it's a conscious decision to merge the upgrade PR or not. Extension developers will know and will have validated this.

Hence, I think we should not persist this and rather default implicitly. This will make operations easier. If you still have concerns, let's have a call about it. cc @timuthy @oliver-goetz @MartinWeindel

Reply:
We can have a group discussion to reach a consensus. With implicit defaulting:
- every time you need to deal with an extension in the codebase, you would need to apply the defaults which is annoying.
- you don't know what is actually being applied, until it is applied. Debugging issues with the values here would not be fun experience.

That being said, the webhook can be used for a "rolling upgrades" scenario. If you apply the webhook on UPDATE requests then a config like: 

```
apiVersion: operator.gardener.cloud/v1alpha1
kind: Extension
metadata:
  name: provider-local
```

will get the updated values when reapplied after an operator upgrade. You get some benefits over the hackathon approach:
- you actually see the values being applied
- you can modify the values easier as an operator 
- you can also control when/if to trigger the update (the reapplication of the yaml would be necessary).

Reply:
The two reasons presented don't seem to justify this approach over the implicit defaulting for me: How often do we deal with an extension in the code base so that applying the defaults gets annoying? "Debugging problems" can be solved with logging the defaulted spec if really needed.

Also, I'm not sure whether it is a good idea to run the webhook in UPDATE. If a human user has defined something in the `spec` of the `Extension` object and we roll out a new `gardener-operator`, it can silently override this without the user noticing.

> you actually see the values being applied

See above, can be solved via logging the `spec`.

> you can modify the values easier as an operator

Easier in the sense that some (probably not all) fields are already "pre-filled"? Maybe, but I'm not sure whether this is a reason outperforming the arguments speaking for implicit defaulting.

> you can also control when/if to trigger the update (the reapplication of the yaml would be necessary).

You can still "pin" the version by explicitly setting it in the `spec` of the `Extension` - then the implicit defaulting should not override it (at least we should be build it this way).